# SQL project

### Table Of Contents <a class="anchor" id="table_of_contents"></a>

* [TASK DESCRIPTION](#task-description)
* [Libraries Used](#Importing-libraries-)
* [Accessing the DB](#Accessing-the-DB-)
* [Observing tables](#Showing-tables-)
* [Missing values](#Checking-for-missing-values-:-)
* [Q1 - Find the number of books released after January 1, 2000.](#Q1-Find-the-number-of-books-released-after-January-1-,-2000.)
* [Q2 - Find the number of user reviews and the average rating for each book](#Q2-Find-the-number-of-user-reviews-and-the-average-rating-for-each-book)
* [Q3 - Identify the publisher that has released the greatest number of books with more than 50 pages](#Q3-Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages)
* [Q4 - Identify the author with the highest average book rating (look only at books with at least 50 ratings)](#Q4-Identify--the-author-with-the-highest-average-book-rating-(-look-only-at-books-with-at-least-50-ratings-))
* [Q5 - Find the average number of text reviews among users who rated more than 50 books](#Q5-Find-the-average-number-of-text--reviews-among-users-who-rated-more-than-50-books)
* [Conclusions](#Conclusions-)


### TASK DESCRIPTION:<a class="anchor" id="chapter1"></a>
You've been given a database of one of book apps for book lovers. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

Task:
- Find the number of books released after January 1, 2000.
- Find the number of user reviews and the average rating for each book.
- Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
- Identify the author with the highest average book rating (look only at books with at least 50 ratings).
- Find the average number of text reviews among users who rated more than 50 books.


### Importing libraries:<a class="anchor" id="chapter2"></a>

In [1]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine

In [2]:
! pip install psycopg2

### Accessing the DB<a class="anchor" id="chapter3"></a>

In [3]:


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

### Showing tables:<a class="anchor" id="chapter4"></a>

In [4]:
tables= ['books', 'authors', 'ratings', 'reviews', 'publishers']
for i in tables:
    print ("The {} table:".format(i))
    query = '''
        SELECT *
        FROM {}
        LIMIT 10
    '''.format(i)
    display(pd.io.sql.read_sql(query, con=engine))

The books table:


book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   
5        6        257               1st to Die (Women's Murder Club  #1)   
6        7        258               2nd Chance (Women's Murder Club  #2)   
7        8        260              4th of July (Women's Murder Club  #4)   
8        9        563                                   A Beautiful Mind   
9       10        445                                 A Bend in the Road   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268  
5        424       2005-05-20           116  
6        400       2005-05-20           116  
7        448       2006-06-01           318  
8        461       2002-02-04           104  
9        341       2005-04-01           116

The authors table:


author_id                                author
0          1                            A.S. Byatt
1          2        Aesop/Laura Harris/Laura Gibbs
2          3                       Agatha Christie
3          4                         Alan Brennert
4          5              Alan Moore/David   Lloyd
5          6                            Alan Paton
6          7           Albert Camus/Justin O'Brien
7          8                         Aldous Huxley
8          9    Aldous Huxley/Christopher Hitchens
9         10  Aleksandr Solzhenitsyn/H.T. Willetts

The ratings table:


rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2
5          6        3  johnsonamanda       4
6          7        3    scotttamara       5
7          8        3    lesliegibbs       5
8          9        4    abbottjames       5
9         10        4   valenciaanne       4

The reviews table:


review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   
5          6        3    lesliegibbs   
6          7        4   valenciaanne   
7          8        4    abbottjames   
8          9        5        npowers   
9         10        5        staylor   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...  
5           Analysis no several cause international.  
6  One there cost another. Say type save. With pe...  
7  Within enough mother. There at system full rec...  
8  Thank now focus realize economy focus fly. Ite...  
9  Game push lot reduce where remember. Including...

The publishers table:


publisher_id                                publisher
0             1                                      Ace
1             2                                 Ace Book
2             3                                Ace Books
3             4                            Ace Hardcover
4             5        Addison Wesley Publishing Company
5             6                                  Aladdin
6             7                       Aladdin Paperbacks
7             8                             Albin Michel
8             9                          Alfred A. Knopf
9            10  Alfred A. Knopf Books for Young Readers

### Checking for missing values:<a class="anchor" id="chapter5"></a>
(none were found)

In [5]:
tables= ['books', 'authors', 'ratings', 'reviews', 'publishers']
for i in tables:
    print ("Missing values for table {}".format(i))
    query = '''
        SELECT *
        FROM {}
        WHERE ({} is null)
        '''.format(i, i)
    display(pd.io.sql.read_sql(query, con=engine))

Missing values for table books


Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

Missing values for table authors


Empty DataFrame
Columns: [author_id, author]
Index: []

Missing values for table ratings


Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

Missing values for table reviews


Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

Missing values for table publishers


Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

## Q1 - Find the number of books released after January 1, 2000 <a class="anchor" id="chapter6"></a>

In [6]:
query= """
    SELECT COUNT(book_id)
    FROM books
    WHERE publication_date > '2000-01-01';
    """

In [7]:
table= pd.io.sql.read_sql(query, con = engine)
table

count
0    819

**# After January 1, 2000 there were 819 books released**.

## Q2 - Find the number of user reviews and the average rating for each book <a class="anchor" id="chapter7"></a>


In [8]:
query= '''
    WITH
    c AS (SELECT book_id, 
                 COUNT(review_id) as review_count
              FROM reviews
              GROUP BY book_id),
              
    a AS (SELECT book_id,
                AVG(rating) as average_rating
                FROM ratings
                GROUP BY book_id),
                
   t AS (SELECT book_id,
                title
            FROM books
            GROUP BY book_id) 
            
                
    SELECT t.title,
            c.review_count,
            a.average_rating
    FROM t 
        left join a on t.book_id = a.book_id
        left join c on t.book_id= c.book_id
    ORDER BY t.title
    
'''
table= pd.io.sql.read_sql(query, con = engine)
table

title  review_count  \
0                                         'Salem's Lot           2.0   
1                   1 000 Places to See Before You Die           1.0   
2    13 Little Blue Envelopes (Little Blue Envelope...           3.0   
3    1491: New Revelations of the Americas Before C...           2.0   
4                                                 1776           4.0   
..                                                 ...           ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)           3.0   
996                        Xenocide (Ender's Saga  #3)           3.0   
997                                    Year of Wonders           4.0   
998                        You Suck (A Love Story  #2)           2.0   
999  Zen and the Art of Motorcycle Maintenance: An ...           4.0   

     average_rating  
0          3.666667  
1          2.500000  
2          4.666667  
3          4.500000  
4          4.000000  
..              ...  
995        3.666667  
996        3.400000  
997        3.200000  
998        4.500000  
999        3.833333  

[1000 rows x 3 columns]

**# The table above shows a list of 1000 book titles, with the number of reviews each one got and their average rating.**

#### - Another option:

In [9]:
query= '''
    SELECT b.title, c.reviews_count, a.average_rating 
    FROM books as b
        LEFT JOIN (SELECT book_id, COUNT(review_id) as reviews_count
                  FROM reviews
                  GROUP BY book_id) as c 
            ON b.book_id = c.book_id 
        LEFT JOIN (SELECT book_id, AVG(rating) as average_rating
                FROM ratings
                GROUP BY book_id) as a
            ON b.book_id = a.book_id
    ORDER BY b.title
'''
question2= pd.io.sql.read_sql(query, con = engine)
question2

title  reviews_count  \
0                                         'Salem's Lot            2.0   
1                   1 000 Places to See Before You Die            1.0   
2    13 Little Blue Envelopes (Little Blue Envelope...            3.0   
3    1491: New Revelations of the Americas Before C...            2.0   
4                                                 1776            4.0   
..                                                 ...            ...   
995           Wyrd Sisters (Discworld  #6; Witches #2)            3.0   
996                        Xenocide (Ender's Saga  #3)            3.0   
997                                    Year of Wonders            4.0   
998                        You Suck (A Love Story  #2)            2.0   
999  Zen and the Art of Motorcycle Maintenance: An ...            4.0   

     average_rating  
0          3.666667  
1          2.500000  
2          4.666667  
3          4.500000  
4          4.000000  
..              ...  
995        3.666667  
996        3.400000  
997        3.200000  
998        4.500000  
999        3.833333  

[1000 rows x 3 columns]

### Q3- Identify the publisher that has released the greatest number of books with more than 50 pages <a class="anchor" id="chapter8"></a>

In [10]:

query= """
    SELECT publishers.publisher,
           COUNT(book_id) AS long_books_count
    FROM publishers left join books on publishers.publisher_id = books.publisher_id
    WHERE books.num_pages > 50 
    GROUP BY publishers.publisher
    ORDER BY long_books_count desc
    LIMIT 1; 
    """

In [11]:
table= pd.io.sql.read_sql(query, con = engine)
table

publisher  long_books_count
0  Penguin Books                42

**# Out of all the publishers that published books with over 50 pages, Penguin Books published the greatest number of books (42).**

### Q4-  Identify the author with the highest average book rating (look only at books with at least 50 ratings) <a class="anchor" id="chapter9"></a>

In [12]:

query= """
   SELECT
        authors.author,
        AVG(ratings.rating) as avg_rating
   FROM books 
       RIGHT JOIN ratings ON books.book_id= ratings.book_id
       RIGHT JOIN authors ON books.author_id= authors.author_id
   GROUP BY authors.author
   HAVING COUNT(ratings.rating) > 50
   ORDER BY avg_rating DESC
   LIMIT 1;
 """

In [13]:
table= pd.io.sql.read_sql(query, con = engine)
table

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.288462

**# Out of all the authors that published books with over 50 pages, .K. Rowling/Mary GrandPré had the highest average book rating (4.29).**

### Q5 - Find the average number of text reviews among users who rated more than 50 books <a class="anchor" id="chapter10"></a>

In [14]:

query= """
SELECT AVG(text_counter.text_count) as avg_text_reviews
FROM (SELECT r.username,
            count(text) as text_count
    FROM reviews as r
    WHERE r.username IN (SELECT username
                        FROM ratings 
                        group by username
                        HAVING count(rating) > 50)
    GROUP BY r.username) AS text_counter;
    """

In [15]:
table= pd.io.sql.read_sql(query, con = engine)
table

avg_text_reviews
0         24.333333

**# The average number of text reviews among users who rated more than 50 books is 24.33.**

## Conclusions: <a class="anchor" id="chapter11"></a>
As part of an initial data-exploration stage, in the process of building an app for book lovers, we wanted to answer a few questions on the book market.
We found that:
- 1- 819 books were released after January 1, 2000.
- 2- Books get reviews (between 0-7 per book) and rating scores (on average between 1.5 to 5).
- 3- Out of all the publishers that published books with over 50 pages, Penguin Books published the greatest number of books (42).
- 4-  Out of all the authors that published books with over 50 pages, .K. Rowling/Mary GrandPré had the highest average book rating (4.29).
- 5- The average number of text reviews among users who rated more than 50 books is 24.33.